src/modules/adaptors/Morpho/MorphoAaveV3DebtTokenAdaptor.sol

```diff
@@ -0,0 +1,156 @@
+     */
+    function HFMIN() internal pure returns (uint256) {
+        return 1.05e18;
+    }
+
+    //============================================ Implement Base Functions ===========================================
+
+    /**
+     * @notice User deposits are NOT allowed into this position.
+     */
+    function deposit(uint256, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserDepositsNotAllowed();
+    }
+
+    /**
+     * @notice User withdraws are NOT allowed from this position.
+     */
+    function withdraw(uint256, address, bytes memory, bytes memory) public pure override {
+        revert BaseAdaptor__UserWithdrawsNotAllowed();
+    }
+
+    /**
+     * @notice This position is a debt position, and user withdraws are not allowed so
+     *         this position must return 0 for withdrawableFrom.
+     */
+    function withdrawableFrom(bytes memory, bytes memory) public pure override returns (uint256) {
+        return 0;
+    }
+
+    /**
+     * @notice Returns the cellars balance of the positions debt.
+     */
+    function balanceOf(bytes memory adaptorData) public view override returns (uint256) {
+        address underlying = abi.decode(adaptorData, (address));
+        return morpho().borrowBalance(underlying, msg.sender);
+    }
+
+    /**
+     * @notice Returns the positions underlying asset.
+     */
+    function assetOf(bytes memory adaptorData) public pure override returns (ERC20) {
+        ERC20 underlying = abi.decode(adaptorData, (ERC20));
+        return underlying;
+    }
+
+    /**
+     * @notice This adaptor reports values in terms of debt.
+     */
+    function isDebt() public pure override returns (bool) {
+        return true;
+    }
+
+    //============================================ Strategist Functions ===========================================
+
+    /**
+     * @notice Allows strategists to borrow assets from Morpho.
+     * @notice `debtTokenToBorrow` must be the debtToken, NOT the underlying ERC20.
+     * @param underlying the debtToken to borrow on Morpho
+     * @param amountToBorrow the amount of `debtTokenToBorrow` to borrow on Morpho.
+     */
+    function borrowFromAaveV3Morpho(address underlying, uint256 amountToBorrow, uint256 maxIterations) public {
+        // Check that debt position is properly set up to be tracked in the Cellar.
+        bytes32 positionHash = keccak256(abi.encode(identifier(), true, abi.encode(underlying)));
+        uint32 positionId = Cellar(address(this)).registry().getPositionHashToPositionId(positionHash);
+        if (!Cellar(address(this)).isPositionUsed(positionId))
+            revert MorphoAaveV3DebtTokenAdaptor__DebtPositionsMustBeTracked(underlying);
+
+        // Borrow from morpho.
+        morpho().borrow(underlying, amountToBorrow, address(this), address(this), maxIterations);
+
+        // Check that health factor is above adaptor minimum.
+        uint256 healthFactor = _getUserHealthFactor(address(this));
```

Comment:
Interesting, does Morpho contracts make it so that you cannot check the health factor as a preview prior to executing `morpho.borrow()`? If they do provide it, then moving this before the `borrow()` would be more gas efficient no?

Reply:
Yeah I think the easiest way to see if we can do a borrow is to actually try doing the borrow, then it will either revert in the morpho contracts(if HF is < 1), or in our adaptor if HF is less than HFMIN. 

There isn't a straight forward function we can call to check to see how a new borrow would impact our health factor, and even if there was, we'd end up spending more gas running the calculation everytime we borrowed, vs the gas we would save from reverting early for the odd times where we borrowed too much.

Reply:
Ah fair enough, yeah it's fine-tuning regarding the conditional logic of checking vs the execution call where we borrowed and had to revert.